### PR TITLE
support remote KUBECONFIG for istiod

### PR DESCRIPTION
**Pull request by irisdingbj**
https://github.com/istio/istio/issues/24897

User need to run `kubectl create secret generic istio-kubeconfig --from-file=config=xxxx` to create the secret first.

### CODE REVIEW

**Review by howardjohn — CHANGES_REQUESTED** [review]
I don't think we should implement it this way. We can already add generic environment variables (kubeconfig flag and KUBECONFIG variable are the same ) and I think we can add arbitrary file mounts. If not we can add the file mounts.

**Comment by irisdingbj** [comment]
> I don't think we should implement it this way. We can already add generic environment variables (kubeconfig flag and KUBECONFIG variable are the same ) and I think we can add arbitrary file mounts. If not we can add the file mounts.

@howardjohn we can set using KUBECONFIG env var or use --kubeconfig command line option as what the PR did.   The problem is user first needs to create a generic secret from existing kubeconfig files.  If we want to use the config data directly like : 
```
valueFrom:
              secretKeyRef:
                name: remote-kube-config
                key: config
```
 it means we need to parse the string data - this needs additional logic in current code base. As currently we expect it as a file path.

**Comment by howardjohn** [comment]
Yes we should just have a generic option for file mounts though - see the
gateway templates which already have this

**Comment by irisdingbj** [comment]
@howardjohn remove the flag as current code base already can take care of no exist kubeconfig . PTAL.

**Comment by linsun** [comment]
@irisdingbj maybe I'm missing something, it is unclear to me where we check if the path/file exists, if so, use the kubeconfig, else, use the default in cluster one.

**Comment by howardjohn** [comment]
Its still unclear to me that hardcoding this is the best option. Why do you prefer this solution rather than allowing generic volume mounts onto the pilot pod and setting this in the profile?

**Comment by irisdingbj** [comment]
> @irisdingbj maybe I'm missing something, it is unclear to me where we check if the path/file exists, if so, use the kubeconfig, else, use the default in cluster one.

Current code base already takes care of no exist kubeconfig, if it does not exist, will fall back to use in-cluster kubeconfig:

https://github.com/istio/istio/blob/master/pkg/kube/util.go#L44

**Comment by irisdingbj** [comment]
> Its still unclear to me that hardcoding this is the best option. Why do you prefer this solution rather than allowing generic volume mounts onto the pilot pod and setting this in the profile?

@howardjohn I am not a fan for hardcoding as well. So I am glad if I can make it more flexible.  Can you give more details of the generic option you suggested? eg, for end user, what profile change should I make? What file mount option should I add into the istiod yaml?  If you can share a snippet for the IOP it will be helpful for me to implement it.

**Comment by linsun** [comment]
@irisdingbj thanks for the link!

@howardjohn I think I prefer preconfigured path.  This is not something I'd rather configure in profile.  What value would I gain to have it configurable in profile?

BTW, isn't this same approach as how we handle plugin your own key/certs vs self generated key/certs using secret cacerts which is also on pre-configured path?

**Comment by howardjohn** [comment]
Having a hardcoded secret name seems a bit suspicious to me but maybe its not so bad - as you mentioned `cacerts` is doing this. However, the name is not a good one in my opinion, `remote-kube-config` is too generic and specific at the same time in my opinion. Maybe `istio-kubeconfig` would be better.

Any thoughts @costinm @nmittler @sdake ?

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml [hunk 1]
```diff
@@ -176,6 +178,9 @@ spec:
           - name: cacerts
             mountPath: /etc/cacerts
             readOnly: true
+          - name: remote-kube-config
+            mountPath: /etc/remote/config
```
Comment (sdake) [inline manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:182]: please change mountPath to /var/run/secrets/remote to be consistent with the rest of the mountPaths in this file.

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml [hunk 2]
```diff
@@ -203,6 +208,10 @@ spec:
         secret:
           secretName: cacerts
           optional: true
+      - name: remote-kube-config
```
Comment (sdake) [inline manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:211]: perhaps `central-kube-config`?

**Review by sdake — COMMENTED** [review]
@howardjohn I don't have a strong preference for the environment variable name, although I do agree that remote-kube-config could be misunderstood.

**Comment by sdake** [comment]
tend to agree we should not add a generic volume mount field to profiles. I could see that becoming confusing for end users. Also, users may end up importing things into pilot the Istio Authors did not intend, causing support issues.

Cheers,
-steve

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml (hunk 2 above)
Comment (irisdingbj) [reply manifests/charts/istio-control/istio-discovery/templates/deployment.yaml:211]: `istio-kubeconfig` might be better as this is a way to define kubeconfig for istiod and is not limited to central mode.

**Comment by irisdingbj** [comment]
@howardjohn @sdake @linsun Updated according to your comments, PTAL

File: manifests/charts/istio-control/istio-discovery/templates/deployment.yaml (hunk 2 above)
Comment (linsun) [reply manifests/charts/istio-control/istio-discovery/templates/deployment.yaml]: +1 on `istio-kubeconfig` - fits different scenarios.

File: manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml [hunk 3]
```diff
@@ -814,6 +816,9 @@ spec:
           - name: cacerts
             mountPath: /etc/cacerts
             readOnly: true
+          - name: istio-kubeconfig
+            mountPath: /var/run/secrets/remote
```
Comment (linsun) [inline manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml:820]: do we keep remote here?  /var/run/secrets/kube

**Review by linsun — APPROVED** [review]
LGTM

@howardjohn PTAL - I think this is good enough for merge.

**Comment by howardjohn** [comment]
Shouldn't the central istiod test be using this?

**Comment by irisdingbj** [comment]
> Shouldn't the central istiod test be using this?

central istio test is based on previous central model that istiod is using local cluste's KUBECONFIG.

**Comment by howardjohn** [comment]
The central istiod test should be testing central istiod. If central istiod is using remote kube config, we should test remote kube config. If it's not then what is this pr for?

**Comment by linsun** [comment]
Yes, central istiod single cluster test should be using this.  We would need to have a test for it.
- central istiod single cluster (using kubeconfig to point to remote cluster)
- central istiod multiclusters 

cc @GregHanson

**Comment by howardjohn** [comment]
I am perfectly fine with having multiple workload clusters being out of scope for this pr, but seems the existing test should move config reading from control plane cluster to workload cluster

**Comment by linsun** [comment]
I'll let @GregHanson weigh in on whether it is easier to update existing test vs creating a new test for it as there are value to keep the multicluster central istiod tests as well.

I don't think this PR should be blocked for merging.   The test can come in as a follow up PR given this feature is dark launched anyway.

**Comment by GregHanson** [comment]
@linsun @howardjohn the existing central istio tests, specifically the shared mutlicluster tests they utilize, are not set up to handle the central istio single cluster scenario.  I recommend starting with a new test and leave the existing one for now

**Review by howardjohn — APPROVED** [review]

**Pull request merged**